runtime/src/address_map/pending_changes.rs

```diff
@@ -0,0 +1,184 @@
+use {
+    dashmap::{
+        mapref::multiple::RefMulti as MapEntry, setref::multiple::RefMulti as SetEntry, DashMap,
+        DashSet,
+    },
+    log::warn,
+    solana_address_map_program::{AddressMapError, AddressMapState},
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::sync::Arc,
+};
+
+/// Address map activations and deactivations that occurred in a single slot and
+/// will be applied to the address map cache when the bank has been rooted.
+#[derive(Debug)]
+pub struct AddressMapPendingChanges {
+    slot: Slot,
+    activations: DashMap<Pubkey, Arc<Vec<Pubkey>>>,
+    deactivations: DashSet<Pubkey>,
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+impl ::solana_frozen_abi::abi_example::AbiExample for AddressMapPendingChanges {
+    fn example() -> Self {
+        AddressMapPendingChanges::new(0)
+    }
+}
+
+impl AddressMapPendingChanges {
+    /// Create new pending changes collection for a slot
+    pub fn new(slot: Slot) -> Self {
+        Self {
+            slot,
+            activations: DashMap::default(),
+            deactivations: DashSet::default(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_for_tests(
+        activations: DashMap<Pubkey, Arc<Vec<Pubkey>>>,
+        deactivations: DashSet<Pubkey>,
+    ) -> Self {
+        Self {
+            slot: 0,
+            activations,
+            deactivations,
+        }
+    }
+
+    /// Iterator over all address maps activated in a single bank
+    pub fn activations_iter(&self) -> impl Iterator<Item = MapEntry<Pubkey, Arc<Vec<Pubkey>>>> {
+        self.activations.iter()
+    }
+
+    /// Iterator over all address maps deactivated in a single bank
+    pub fn deactivations_iter(&self) -> impl Iterator<Item = SetEntry<Pubkey>> {
+        self.deactivations.iter()
+    }
+
+    /// Record pending change for address map account if it was newly activated or deactivated
+    pub fn record(&self, key: &Pubkey, map_data: &[u8]) {
```

Comment:
Seems like this one could also be fallible

Reply:
Updated to return a result